### PR TITLE
sharpen: optimize code and remove SSE

### PR DESCRIPTION
**Pull request by ralfbrown**
The compiler was producing code that was just as fast as the hand-written SSE, so remove the SSE codepath to reduce future maintenance.  Then, optimize the plain code by using the same rearrangement of horizontal and vertical passes used in other iops.  Finally, merge the subtraction step into the blur loop to avoid a second pass over the output buffer and reduce memory traffic (helps at high thread counts).

Timings with integration test 0028 on a 32-core Threadripper:
```
Thr   Old     New
1     0.127   0.103
4     0.035   0.026
8     0.021   0.013
16   0.015   0.008
24   0.016   0.007
32   0.016   0.007
```

### CODE REVIEW

File: src/iop/sharpen.c [hunk 1]
```diff
@@ -275,388 +277,144 @@ void tiling_callback(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t
...
-// gauss blur the image horizontally
+  const float *const restrict in = (float*)ivoid;
+  const size_t width = roi_out->width;
 #ifdef _OPENMP
 #pragma omp parallel for default(none) \
-  dt_omp_firstprivate(ch, ivoid, mat, rad, roi_in, roi_out, tmp, wd4) \
+  dt_omp_firstprivate(in, ovoid, mat, rad, width, roi_in, roi_out, tmp, padded_size) \
   schedule(static)
 #endif
+//  for(int j = rad; j < roi_out->height - rad; j++)
```
Comment (TurboGit) [inline src/iop/sharpen.c:346]: probably to be removed, a leftovert from debug?

File: src/iop/sharpen.c [hunk 2]
```diff
@@ -275,388 +277,144 @@ void tiling_callback(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t
...
-      const float *inp = in - step * rad;
-      const float *m = mat;
       float sum = 0.0f;
-      for(int k = -rad; k <= rad; k++, m++, inp += step) sum += *m * *inp;
-      *out = sum;
-      out += ch;
-      in++;
+      for(int k = i-rad; k <= i+rad; k++)
+      {
+        int k_adj = k - (i-rad);
```
Comment (TurboGit) [inline src/iop/sharpen.c:401]: const

File: src/iop/sharpen.c [hunk 3]
```diff
@@ -275,388 +277,144 @@ void tiling_callback(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t
...
-      const float *inp = in - step * rad;
-      __m128 msum = _mm_setzero_ps();
-
-      for(int k = 0; k < wd4 * 4; k += 4, inp += step * 4)
+      float sum = 0.0f;
+      for(int k = start_row; k <= end_row; k++)
       {
-        msum = _mm_add_ps(msum, _mm_mul_ps(_mm_load_ps(mat + k),
-                                           _mm_set_ps(inp[3 * step], inp[2 * step], inp[step], inp[0])));
+        int k_adj = k - (j-rad);
```
Comment (TurboGit) [inline src/iop/sharpen.c:385]: const

File: src/iop/sharpen.c [hunk 4]
```diff
@@ -275,388 +277,144 @@ void tiling_callback(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t
...
     {
-      const float *inp = in - ch * rad;
-      const float *m = mat;
-      float sum = 0.0f;
-      for(int k = -rad; k <= rad; k++, m++, inp += ch)
+      float DT_ALIGNED_PIXEL sum[4] = { 0.0f };
+      for(int k = start_row; k <= end_row; k++)
       {
-        sum += *m * *inp;
+        int k_adj = k - (j-rad);
```
Comment (TurboGit) [inline src/iop/sharpen.c:371]: const

File: src/iop/sharpen.c [hunk 5]
```diff
@@ -275,388 +277,144 @@ void tiling_callback(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t
...
+      for(size_t i = 0; i < width; i++)
+        row_out[4*i] = row_in[4*i];
+      continue;
     }
-    for(; i < roi_out->width - rad; i++)
+    // Get a thread-local temporary buffer for processing the current row of the image.
+    float *const restrict temp_buf = dt_get_perthread(tmp, padded_size);
+    // vertically blur the pixels of the current row into the temp buffer
+    size_t start_row = j-rad;
+    size_t end_row = j+rad;
```
Comment (TurboGit) [inline src/iop/sharpen.c:364]: const

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some minor changes. I have tested again the integration test, all good.

File: src/iop/sharpen.c (hunk 1 above)
Comment (ralfbrown) [reply src/iop/sharpen.c]: Yup

**Review by TurboGit — APPROVED** [review]
Thanks !

**Pull request merged**